// bmad-speckit/src/services/template-fetcher.js
/**
 * TemplateFetcher - minimal implementation (GAP-4.1, plan §3.1)
 * Story 11.1: cache to ~/.bmad-speckit/templates/<template-id>/<tag>/, --template tag|url, networkTimeoutMs.
 * templateSource: opts.templateSource || env SDD_TEMPLATE_REPO || 'bmad-method/bmad-method'
 */
const fs = require('fs');
const path = require('path');
const os = require('os');
const https = require('https');
const http = require('http');
const { createGunzip } = require('zlib');

const DEFAULT_TEMPLATE_REPO = 'bmad-method/bmad-method';
const DEFAULT_NETWORK_TIMEOUT_MS = 30000;

function getCacheRoot() {
  return path.join(os.homedir(), '.bmad-speckit', 'templates');
}

function _getTemplateSource(opts) {
  return opts?.templateSource || process.env.SDD_TEMPLATE_REPO || DEFAULT_TEMPLATE_REPO;
}

function getTemplateId(templateSource) {
  const src = templateSource || process.env.SDD_TEMPLATE_REPO || DEFAULT_TEMPLATE_REPO;
  const parts = src.split('/');
  return parts.length > 1 ? parts[1] : 'default';
}

function getCacheDir(tag, templateSource) {
  const subdir = tag === 'latest' ? 'latest' : tag;
  const tid = templateSource != null ? getTemplateId(templateSource) : getTemplateId();
  return path.join(getCacheRoot(), tid, subdir);
}

function isCacheValid(cacheDir) {
  if (!fs.existsSync(cacheDir) || !fs.statSync(cacheDir).isDirectory()) return false;
  const entries = fs.readdirSync(cacheDir);
  return entries.length > 0;
}

/**
 * Check for local _bmad in monorepo (packages/bmad-speckit/src/services -> repo root)
 * Returns repo root path (parent of _bmad) so generateSkeleton finds _bmad and _bmad-output
 */
function getLocalTemplatePath() {
  const repoRoot = path.join(__dirname, '../../../..');
  const bmadPath = path.join(repoRoot, '_bmad');
  if (fs.existsSync(bmadPath) && fs.statSync(bmadPath).isDirectory()) {
    const entries = fs.readdirSync(bmadPath);
    if (entries.length > 0) return repoRoot;
  }
  return null;
}

/**
 * Fetch tarball from GitHub, extract to cache dir, return path.
 * On failure: throw with message (caller exits 3).
 * Story 11.1: cache to ~/.bmad-speckit/templates/<template-id>/<tag|latest>/; reuse if valid.
 */
async function fetchFromGitHub(tag, opts = {}) {
  const { githubToken, skipTls, debug, networkTimeoutMs } = opts;
  const templateSource = _getTemplateSource(opts);
  const timeoutMs = networkTimeoutMs ?? parseInt(process.env.SDD_NETWORK_TIMEOUT_MS || '', 10) || DEFAULT_NETWORK_TIMEOUT_MS;
  if (skipTls) {
    console.warn('Warning: --skip-tls skips SSL/TLS verification. Use only in trusted networks.');
  }

  const cacheDir = getCacheDir(tag, templateSource);
  if (isCacheValid(cacheDir)) {
    if (debug) console.error('[bmad-speckit:debug] Using cache:', cacheDir);
    return cacheDir;
  }

  const [owner, repo] = templateSource.split('/');
  let tagToUse = tag;
  if (tag === 'latest') {
    const apiUrl = `https://api.github.com/repos/${owner}/${repo}/releases/latest`;
    const headers = { 'User-Agent': 'bmad-speckit' };
    if (githubToken) headers.Authorization = `token ${githubToken}`;

    const latest = await fetchJson(apiUrl, { headers, skipTls, timeout: timeoutMs });
    if (!latest || !latest.tag_name) {
      const e = new Error('Failed to get latest release from GitHub. Check network or use --github-token.');
      e.code = 'NETWORK_TEMPLATE';
      throw e;
    }
    tagToUse = latest.tag_name;
    if (debug) console.error('[bmad-speckit:debug] latest tag:', tagToUse);
  }

  const tarballUrl = `https://github.com/${owner}/${repo}/archive/refs/tags/${tagToUse}.tar.gz`;
  fs.mkdirSync(cacheDir, { recursive: true });

  try {
    await downloadAndExtract(tarballUrl, cacheDir, { githubToken, skipTls, timeout: timeoutMs });
    return cacheDir;
  } catch (err) {
    if (fs.existsSync(cacheDir)) {
      try {
        fs.rmSync(cacheDir, { recursive: true });
      } catch {}
    }
    const e = new Error(`Template fetch failed: ${err.message}. Use --offline with cached template (Story 11.2).`);
    e.code = 'NETWORK_TEMPLATE';
    throw e;
  }
}

const TIMEOUT_MESSAGE = '网络超时 (Network timeout)';

function fetchJson(url, opts) {
  return new Promise((resolve, reject) => {
    const timeout = opts.timeout ?? DEFAULT_NETWORK_TIMEOUT_MS;
    const lib = url.startsWith('https') ? https : http;
    const req = lib.get(url, { rejectUnauthorized: !opts.skipTls }, (res) => {
      let data = '';
      res.on('data', (c) => (data += c));
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch {
          const e = new Error('Invalid JSON response');
          e.code = 'NETWORK_TEMPLATE';
          reject(e);
        }
      });
    });
    req.on('error', (err) => {
      const e = new Error(err.message);
      e.code = 'NETWORK_TEMPLATE';
      reject(e);
    });
    req.setTimeout(timeout, () => {
      req.destroy();
      const e = new Error(TIMEOUT_MESSAGE);
      e.code = 'NETWORK_TEMPLATE';
      reject(e);
    });
  });
}

function downloadAndExtract(url, destDir, opts) {
  return new Promise((resolve, reject) => {
    const timeout = opts.timeout ?? DEFAULT_NETWORK_TIMEOUT_MS;
    const headers = { 'User-Agent': 'bmad-speckit' };
    if (opts.githubToken) headers.Authorization = `token ${opts.githubToken}`;

    const lib = url.startsWith('https') ? https : http;
    const req = lib.get(url, { rejectUnauthorized: !opts.skipTls, headers }, (res) => {
      if (res.statusCode !== 200) {
        const e = new Error(`模板拉取失败: HTTP ${res.statusCode}`);
        e.code = 'NETWORK_TEMPLATE';
        reject(e);
        return;
      }
      const tar = require('tar');
      const gunzip = createGunzip();
      res.pipe(gunzip).pipe(tar.extract({ cwd: destDir, strip: 1 }))
        .on('finish', resolve)
        .on('error', (err) => {
          const e = new Error(`模板解压失败: ${err.message}`);
          e.code = 'NETWORK_TEMPLATE';
          reject(e);
        });
    });
    req.on('error', (err) => {
      const e = new Error(err.message);
      e.code = 'NETWORK_TEMPLATE';
      reject(e);
    });
    req.setTimeout(timeout, () => {
      req.destroy();
      const e = new Error(TIMEOUT_MESSAGE);
      e.code = 'NETWORK_TEMPLATE';
      reject(e);
    });
  });
}

/**
 * Main export: fetch template, return path to template dir.
 * templateSpec: 'latest' | tag (e.g. 'v1.0.0') | url (https://...).
 */
async function fetchTemplate(templateSpec, opts = {}) {
  const local = getLocalTemplatePath();
  if (local) {
    if (opts.debug) console.error('[bmad-speckit:debug] Using local _bmad:', local);
    return local;
  }
  const trimmed = String(templateSpec || 'latest').trim();
  if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) {
    return fetchFromUrl(trimmed, opts);
  }
  return fetchFromGitHub(trimmed, opts);
}

/**
 * Fetch tarball from custom URL, extract to cache url-<hash>, return path.
 */
async function fetchFromUrl(url, opts = {}) {
  const templateSource = _getTemplateSource(opts);
  const timeoutMs = opts.networkTimeoutMs ?? parseInt(process.env.SDD_NETWORK_TIMEOUT_MS || '', 10) || DEFAULT_NETWORK_TIMEOUT_MS;
  const crypto = require('crypto');
  const hash = crypto.createHash('sha256').update(url).digest('hex').slice(0, 12);
  const tid = getTemplateId(templateSource);
  const cacheDir = path.join(getCacheRoot(), tid, `url-${hash}`);
  fs.mkdirSync(cacheDir, { recursive: true });
  try {
    await downloadAndExtract(url, cacheDir, { timeout: timeoutMs, skipTls: opts.skipTls });
    return cacheDir;
  } catch (err) {
    if (fs.existsSync(cacheDir)) {
      try { fs.rmSync(cacheDir, { recursive: true }); } catch {}
    }
    if (!err.code) {
      err.code = 'NETWORK_TEMPLATE';
    }
    throw err;
  }
}

module.exports = {
  fetchTemplate,
  fetchFromGitHub,
  fetchFromUrl,
  getLocalTemplatePath,
  getCacheRoot,
  getTemplateId,
  getCacheDir,
  isCacheValid,
  TIMEOUT_MESSAGE,
  DEFAULT_NETWORK_TIMEOUT_MS,
};
